<a href="https://colab.research.google.com/github/thewhitedrive/mlinternflyrank/blob/main/work/notebooks/w01_research_question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/thewhitedrive/mlinternflyrank/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*


My Lane: Raanking signal Analysis
I cchose the Ranking Signal analysis lane because the starter dataset contains search performance metrics such as impressions, clicks , CTR and average position , along with content related imformation . There is a huge gap between Mean and Median values across different metrics , suggesting that performance is highly skewed . I would like to investigate which measurable signals are associatd with stronger search performance .

In [2]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

Research questions : which measurable search and content signals are associated with stronger search performance and can they help us prioritze content items for SEO review
DECISION : which content items should be prioritize for SEO review
ACTION : To ensure the content team reviews the highest priority items and take proper action
COST OF A WRONG CALL  : Prioritzing thw wrong content item can waste considerable time amnd resources , while failing to prioritize a valuable item can result in degraded efficiency and performance
  

In [3]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [4]:
import pandas as pd
url = "https://raw.githubusercontent.com/thewhitedrive/mlinternflyrank/main/data/raw/content_refresh_anonymized.csv"
df=pd.read_csv(url)
print(" Shape :",df.shape)
print("\nColumns :")
print(df.columns.tolist())

 Shape : (30000, 44)

Columns :
['content_id', 'client_id', 'search_volume', 'competition', 'competition_level', 'cpc', 'content_type', 'main_intent', 'word_count', 'char_count', 'provider_used', 'model_used', 'impressions_90d', 'clicks_90d', 'pageviews_90d', 'sessions_90d', 'users_90d', 'engaged_sessions_90d', 'ai_sessions_90d', 'scroll_events_90d', 'days_with_impressions', 'days_with_sessions', 'impressions_last_30d', 'clicks_last_30d', 'sessions_last_30d', 'impressions_prev_30d', 'clicks_prev_30d', 'sessions_prev_30d', 'content_age_days', 'age_tier', 'age_tier_order', 'days_since_last_update', 'freshness_tier', 'word_count_tier', 'char_count_tier', 'ctr', 'avg_position', 'engagement_rate', 'scroll_rate', 'ai_traffic_pct', 'impression_tier', 'position_tier', 'trend_direction', 'trend_pct']


In [5]:
df.describe()

,search_volume,competition,cpc,word_count,char_count,impressions_90d,clicks_90d,pageviews_90d,sessions_90d,users_90d,...,sessions_prev_30d,content_age_days,age_tier_order,days_since_last_update,ctr,avg_position,engagement_rate,scroll_rate,ai_traffic_pct,trend_pct
count,27532.000000,27532.000000,27532.000000,22301.000000,22301.000000,30000.000000,30000.000000,30000.000000,30000.000000,30000.000000,...,30000.000000,30000.00000,30000.000000,30000.000000,30000.000000,30000.00000,30000.000000,29875.000000,30000.000000,26612.000000
mean,158.882391,0.146954,0.485342,3107.760325,20665.277835,5200.366300,16.097333,49.942467,37.066633,35.937700,...,10.283000,256.16780,4.786533,46.098300,0.510733,16.34238,2.534520,18.212921,0.768196,-4.785969
std,1518.270825,0.285241,2.101560,1452.382598,10115.344042,16838.019547,75.076958,152.101430,107.069131,103.748185,...,42.578003,132.70793,0.790392,42.078709,3.279162,15.21679,8.310096,29.472768,7.429454,473.861780
min,0.000000,0.000000,0.000000,8.000000,40.000000,1.000000,0.000000,0.000000,1.000000,1.000000,...,0.000000,90.00000,3.000000,1.000000,0.000000,0.00000,0.000000,0.000000,0.000000,-100.000000
25%,0.000000,0.000000,0.000000,2413.000000,15644.000000,81.000000,0.000000,2.000000,2.000000,2.000000,...,1.000000,132.00000,4.000000,20.000000,0.000000,6.20000,0.000000,0.000000,0.000000,-62.600000
50%,10.000000,0.000000,0.000000,2877.000000,19116.000000,731.000000,1.000000,8.000000,7.000000,7.000000,...,2.000000,236.00000,5.000000,20.000000,0.070000,10.80000,0.000000,5.000000,0.000000,-33.500000
75%,20.000000,0.130000,0.000000,3666.000000,24011.000000,3615.250000,7.000000,33.000000,27.000000,27.000000,...,7.000000,333.00000,5.000000,104.000000,0.290000,22.30000,1.350000,23.530000,0.000000,0.000000
max,74000.000000,1.000000,100.360000,9546.000000,111158.000000,517715.000000,4178.000000,5998.000000,4345.000000,4913.000000,...,4247.000000,564.00000,6.000000,373.000000,100.000000,245.00000,100.000000,300.000000,300.000000,44900.000000


In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 44 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   content_id              30000 non-null  object 
 1   client_id               30000 non-null  object 
 2   search_volume           27532 non-null  float64
 3   competition             27532 non-null  float64
 4   competition_level       27390 non-null  object 
 5   cpc                     27532 non-null  float64
 6   content_type            30000 non-null  object 
 7   main_intent             27626 non-null  object 
 8   word_count              22301 non-null  float64
 9   char_count              22301 non-null  float64
 10  provider_used           8562 non-null   object 
 11  model_used              24267 non-null  object 
 12  impressions_90d         30000 non-null  int64  
 13  clicks_90d              30000 non-null  int64  
 14  pageviews_90d           30000 non-null

In [7]:
# Show columns related to ranking, traffic and engagement

important_words = [
    "position",
    "impression",
    "click",
    "ctr",
    "session",
    "engagement",
    "search",
    "pageview"
]

for col in df.columns:
    if any(word in col.lower() for word in important_words):
        print(col)



search_volume
impressions_90d
clicks_90d
pageviews_90d
sessions_90d
engaged_sessions_90d
ai_sessions_90d
days_with_impressions
days_with_sessions
impressions_last_30d
clicks_last_30d
sessions_last_30d
impressions_prev_30d
clicks_prev_30d
sessions_prev_30d
ctr
avg_position
engagement_rate
impression_tier
position_tier


In [8]:
df[
    [col for col in df.columns
     if any(word in col.lower() for word in
            ["position", "impression", "click", "ctr", "engagement"])]
].describe()

,impressions_90d,clicks_90d,days_with_impressions,impressions_last_30d,clicks_last_30d,impressions_prev_30d,clicks_prev_30d,ctr,avg_position,engagement_rate
count,30000.000000,30000.000000,30000.000000,30000.000000,30000.000000,30000.000000,30000.000000,30000.000000,30000.00000,30000.000000
mean,5200.366300,16.097333,61.454033,1429.058733,4.933867,1783.078500,5.435100,0.510733,16.34238,2.534520
std,16838.019547,75.076958,32.689245,5643.852081,23.929393,6150.429511,28.358673,3.279162,15.21679,8.310096
min,1.000000,0.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.00000,0.000000
25%,81.000000,0.000000,31.000000,10.000000,0.000000,19.000000,0.000000,0.000000,6.20000,0.000000
50%,731.000000,1.000000,81.000000,139.000000,0.000000,210.000000,0.000000,0.070000,10.80000,0.000000
75%,3615.250000,7.000000,88.000000,768.000000,2.000000,1143.000000,2.000000,0.290000,22.30000,1.350000
max,517715.000000,4178.000000,88.000000,238796.000000,1176.000000,218786.000000,1627.000000,100.000000,245.00000,100.000000


In [9]:
print("Number of content items:", len(df))

print("Median 90-day clicks:", df["clicks_90d"].median())

print("Median 90-day impressions:", df["impressions_90d"].median())

print("Median average position:", df["avg_position"].median())

Number of content items: 30000
Median 90-day clicks: 1.0
Median 90-day impressions: 731.0
Median average position: 10.8


Quick look at the data

The starter dataset contains 30,000 content items. The median number of clicks over the last 90 days is 1, while the median number of impressions is 731. The median average search position is 10.8. These values suggest that search performance is uneven across the content set, which makes it useful to investigate which measurable signals are associated with stronger-performing content.

In [10]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.

## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

What I can and can’t claim

I can identify patterns and associations between measurable search/content signals and search performance in this dataset. These findings can provide directional evidence and help the SEO/content team prioritize which content items deserve further review. However, the analysis cannot prove that a particular signal causes better search performance, and it cannot guarantee future traffic or predict Google’s ranking decisions.

In [11]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.